obfuscation.go

```diff
@@ -0,0 +1,188 @@
+}
+
+// Obfuscate is used to make data obfuscation.
+// In context of Lightning Network is either used by the nodes in order to
+// make initial obfuscation with the creation of the hmac or by the forwarding
+// nodes for backward failure obfuscation of the onion failure blob. By
+// obfuscating the onion failure on every node in the path we are adding
+// additional step of the security and barrier for malware nodes to retrieve
+// valuable information. The reason for using onion obfuscation is to not give
+// away to the nodes in the payment path the information about the exact failure
+// and its origin.
+func (o *OnionObfuscator) Obfuscate(initial bool, data []byte) []byte {
+	if initial {
+		umKey := generateKey("um", o.sharedSecret)
+		hash := hmac.New(sha256.New, umKey[:])
+		hash.Write(data)
+		h := hash.Sum(nil)
+		data = append(h, data...)
+	}
+
+	return onionObfuscation(o.sharedSecret, data)
+}
+
+// Decode initializes the obfuscator from the byte stream.
+func (o *OnionObfuscator) Decode(r io.Reader) error {
+	_, err := r.Read(o.sharedSecret[:])
+	return err
+}
+
+// Encode writes converted obfuscator in the byte stream.
+func (o *OnionObfuscator) Encode(w io.Writer) error {
+	_, err := w.Write(o.sharedSecret[:])
+	return err
+}
+
+// OnionDeobfuscator represents the serializable object which encapsulate the
+// all necessary data to properly de-obfuscate previously obfuscated data.
+// In context of Lightning Network the data which have to be deobfuscated
+// usually is onion failure.
+type OnionDeobfuscator struct {
+	sessionKey  *btcec.PrivateKey
+	paymentPath []*btcec.PublicKey
+}
+
+// NewOnionDeobfuscator creates new instance of onion deobfuscator.
+func NewOnionDeobfuscator(sessionKey *btcec.PrivateKey,
+	paymentPath []*btcec.PublicKey) (*OnionDeobfuscator, error) {
+	return &OnionDeobfuscator{
+		sessionKey:  sessionKey,
+		paymentPath: paymentPath,
+	}, nil
+}
+
+// Deobfuscate makes data deobfuscation. The onion failure is obfuscated in
+// backward manner, starting from the node where error have occurred, so in
+// order to deobfuscate the error we need get all shared secret and apply
+// obfuscation in reverse order.
+func (o *OnionDeobfuscator) Deobfuscate(obfuscatedData []byte) (*btcec.PublicKey,
+	[]byte, error) {
+	for i, sharedSecret := range generateSharedSecrets(o.paymentPath,
+		o.sessionKey) {
+		obfuscatedData = onionObfuscation(sharedSecret, obfuscatedData)
+		umKey := generateKey("um", sharedSecret)
+
+		// Split the data and hmac.
+		expectedMac := obfuscatedData[:sha256.Size]
+		data := obfuscatedData[sha256.Size:]
+
+		// Calculate the real hmac.
+		h := hmac.New(sha256.New, umKey[:])
+		h.Write(data)
+		realMac := h.Sum(nil)
+
+		if bytes.Equal(realMac, expectedMac) {
+			return o.paymentPath[i], data, nil
+		}
+	}
+
+	return nil, nil, errors.New("unable to retrieve onion failure")
+}
+
+// Decode initializes the deobfuscator from the byte stream.
+func (o *OnionDeobfuscator) Decode(r io.Reader) error {
```

Comment:
Are these encode and decode methods really necessary? Where are they used? 

I guess we may use them within `lnd` later when we start to persist the raw circuit information as well? In that case they should be removed to the to-be-create `sphinx.Circuit` struct.